modules/xmpp/XmppConnection.js

```diff
@@ -433,25 +433,53 @@ export default class XmppConnection extends Listenable {
         const url = websocketKeepAliveUrl ? websocketKeepAliveUrl
             : this.service.replace('wss://', 'https://').replace('ws://', 'http://');
 
-        return fetch(url)
-            .then(response => {
-
-                // skips header checking if there is no info in options
-                if (!shard) {
-                    return;
-                }
-
-                const responseShard = response.headers.get('x-jitsi-shard');
-
-                if (responseShard !== shard) {
-                    logger.error(
-                        `Detected that shard changed from ${shard} to ${responseShard}`);
-                    this.eventEmitter.emit(XmppConnection.Events.CONN_SHARD_CHANGED);
-                }
-            })
+        // #bloomberg #shard @rpang27 use prosody url to get shard
+        
+        // return fetch(url)
+        //     .then(response => {
+        //         // skips header checking if there is no info in options
+        //         if (!shard) {
+        //             return;
+        //         }
+                
+        //         const responseShard = response.headers.get('x-jitsi-shard');
+                
+        //         if (responseShard !== shard) {
+        //             logger.error(
+        //                 `Detected that shard changed from ${shard} to ${responseShard}`);
+        //                 this.eventEmitter.emit(XmppConnection.Events.CONN_SHARD_CHANGED);
+        //             }
+        //     })
+        //     .catch(error => {
+        //         logger.error(`Websocket Keep alive failed for url: ${url}`, { error });
+        //     });
+        
+        const getShardUrl = this.service.replace('wss://', 'https://').replace('ws://', 'http://').replace('xmpp-websocket', 'shard');
+        
+        const promise = fetch(url)
```

Comment:
I liked the idea of capturing a single operation in a promise. As keep alive and shard detection are separate operations that do not depend on each other we could do a [Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled).

Something like -
```js
// const keepAlivePromise = <fetch with then and catch clauses>;
// const shardDetectionPromise = <another fetch with then and catch clauses>;

if (!shard) {
    return keepAlivePromise;
} else {
    return Promise.allSettled([keepAlivePromise, shardDetectionPromise]);
}
```

Reply:
brilliant idea! changed